src/query_graph/mod.rs

```diff
@@ -382,4 +386,126 @@ impl QueryGraph {
     pub(crate) fn non_trivial_followup_edges(&self) -> &IndexMap<EdgeIndex, IndexSet<EdgeIndex>> {
         &self.non_trivial_followup_edges
     }
+
+    pub(crate) fn edge_for_field(
+        &self,
+        node: NodeIndex,
+        field: &NormalizedField,
+    ) -> Result<Option<EdgeIndex>, FederationError> {
+        let mut candidates = vec![];
+        for edge_ref in self.graph.edges_directed(node, Direction::Outgoing) {
+            let edge_weight = edge_ref.weight();
+            let QueryGraphEdgeTransition::FieldCollection {
+                field_definition_position,
+                ..
+            } = &edge_weight.transition
+            else {
+                continue;
+            };
+            if field.data().field_position == *field_definition_position {
+                candidates.push(edge_ref.id())
+            }
+        }
+        if candidates.len() > 1 {
+            Err(SingleFederationError::Internal {
+                message: format!("Node has multiple edges matching {}", field),
+            }
+            .into())
```

Comment:
Is it possible to end up with many candidates? Instead of finding them all could the loop end early as soon as finding the second one? (And keep an `Option` local variable instead of a `Vec`.)

Reply:
The JS codebase was doing the check here. It shouldn't be possible to have multiple candidates (the progressive override work we'll have to eventually port will change the story a bit, but ultimately it's still only one candidate expected). I'm down with assuming validity here, and just returning early instead of checking the rest of the `Vec` for another match.